### ISSUE TITLE
DataCollatorForCompletionOnlyLM defaults can create a model that never generates `eos_token_id`

### ISSUE THREAD

**Issue by rsnm2** [huggingface/trl#976]
Copied from: https://github.com/huggingface/trl/pull/456#issuecomment-1804665261

First of all - thanks for the great library

I have been trying to fine-tune some models using this collator, but the resulting model I created does not generate the stop token `</s>`. I have been using Mistral, which of course does not have a `pad_token_id`, so I set it to `eos_token_id`

Looking into the implementation `DataCollatorForCompletionOnlyLM`, I see that it inherits from `CollatorForLanguageModeling`, which sets label=-100 for all the pad tokens. Since we set the pad_token_id=eos_token_id, this means that the eos_token is getting a label of -100, meaning we never teach the model to end a sentence :). When I have been using the model I trained with this flow, I have been seeing that the model rarely (if ever) generates the `eos_token`. I wanted to flag this b/c I am sure many will face the same issue

---

Is it possible to set `tokenizer.pad_token_id=tokenizer.bos_token_id` as a workaround?

- In concept, not backpropogating loss for `bos_token_id` seems okay (and I don't think there is any case where we ever end up predicting this token since it should only ever be the first token in the sequence with `attention_mask=1`). However, I am not sure if I don't know what I don't know

**Comment by younesbelkada** [comment]
cc @tomaarsen can you elaborate more on how you subclassed `DataCollatorForLanguageModeling` and fixed your issue? 🙏

**Comment by MustSave** [comment]
In my case, i used the vicuna's training solution which set pad token to unk token

**Comment by younesbelkada** [comment]
Thanks a lot @tomaarsen for sharing your solution
Thanks also @MustSave for your solution as well

@tomaarsen - if I am not mistaken your solution would work well in case one uses `packing` (i.e. no padding token in the input sequences), for that case @MustSave 's solution would be better but I am not sure if all tokenizers have an `unk_token` attribute.

**Comment by tomaarsen** [comment]
So it would still not be completely foolproof in that scenario, I imagine. Definitely tricky.

**Comment by rsnm2** [comment]
thanks for the feedback --> I think using a custom collator that sets `label[i]=-100` if `attention_mask[i]=0` is the approach to take

**Comment by luffycodes** [comment]
Sorry for the question, but I am also facing this issue. is there a temporary solution that might work for mistral (if one is using packing or without packing)?

**Comment by JohnGiorgi** [comment]
> Sorry for the question, but I am also facing this issue. is there a temporary solution that might work for mistral (if one is using packing or without packing)?

Something that has worked for me is to inspect the model and see if it has an unused special token I can use as padding. E.g. for [Llama-3.1-8B](https://huggingface.co/meta-llama/Llama-3.1-8B), setting `pad_token_id=128002` seems to work fine (corresponds to `"<|reserved_special_token_0|>"`).
